### DISCUSSION TITLE
Create product/variants when calculating a price for a delivery

### DISCUSSION THREAD

**Discussion by alexsegura** [coopcycle/coopcycle-web#4169] [General]
The goal here is to improve a little bit the understanding of how the price was calculated. 

Currently, when we calculate the price for a delivery, the resulting order only contains **ONE** item, pointing to a variant of the "On Demand Delivery" product (it is [created by the setup command](https://github.com/coopcycle/coopcycle-web/blob/ca4f5df9ef37c5b0abcad6d695abc9e0ff9247b2/src/Command/SetupCommand.php#L321-L352)). This variant is named "Delivery 3km" (for example). 

So, we lose track of which rules were matched, we only have the resulting price. 

To improve this, we will create a variant for each matched rule, with a human readable name (depending on the rules, it may be a long name, but at least it will be understandable). 

Some code needs to be refactored to implement this. 

- The [`DeliveyManager::getPrice`](https://github.com/coopcycle/coopcycle-web/blob/ca4f5df9ef37c5b0abcad6d695abc9e0ff9247b2/src/Service/DeliveryManager.php#L47-L85) method will need to be deprecated, in favor of another method that implements a sort of "visitor" pattern, to return a structure with the matched rules, and the corresponding price, instead of an integer. 
- Probably most of the logic should be moved into [`OrderFactory::createForDelivery`](https://github.com/coopcycle/coopcycle-web/blob/ca4f5df9ef37c5b0abcad6d695abc9e0ff9247b2/src/Sylius/Order/OrderFactory.php#L87-L126)

**Comment by Paul-Eraman-CoopCycle** [comment]
can you explain this more @alexsegura ?
@r0xsh what was your concern?

**Comment by Paul-Eraman-CoopCycle** [comment]
@Paul-Eraman-CoopCycle can you (i) explain where this information should be shown to 

- [ ] clients (admin of stores)
- [ ] administrators of cooperatives
- [ ] others?

**Comment by alexsegura** [comment]
Work has been started in the [pricing-rule-visitor branch](https://github.com/coopcycle/coopcycle-web/compare/pricing-rule-visitor)

**Comment by Paul-Eraman-CoopCycle** [comment]
what would it look like to use products instead of variants?

**Comment by Paul-Eraman-CoopCycle** [comment]
We could call this:
1. Automatically converting pricing rules into human readable text feature - M
2. Let coops make descriptions of their pricing rule elements feature - M 

and make 1 the backup option if the cooperative has not made a description
